src/dragonball/src/vm/mod.rs

```diff
@@ -89,8 +144,659 @@ impl Default for VmConfigInfo {
+            .as_mut()
+            .ok_or(StartMicroVmError::MissingKernelConfig)?;
+
+        info!(self.logger, "VM: create interrupt manager");
+        self.device_manager
+            .create_interrupt_manager()
+            .map_err(StartMicroVmError::DeviceManager)?;
+
+        info!(self.logger, "VM: create devices");
+        let vm_as =
+            self.address_space
+                .get_vm_as()
+                .ok_or(StartMicroVmError::AddressManagerError(
+                    AddressManagerError::GuestMemoryNotInitialized,
+                ))?;
+        self.device_manager.create_devices(
+            vm_as.clone(),
+            epoll_manager,
+            kernel_config,
+            com1_sock_path,
+            self.dmesg_fifo.take(),
+            self.address_space.address_space(),
+        )?;
+
+        info!(self.logger, "VM: start devices");
+        self.device_manager.start_devices()?;
+
+        info!(self.logger, "VM: initializing devices done");
+        Ok(())
+    }
+
+    /// Remove devices when shutdown vm
+    pub fn remove_devices(&mut self) -> std::result::Result<(), StopMicrovmError> {
+        info!(self.logger, "VM: remove devices");
+        let vm_as = self
+            .address_space
+            .get_vm_as()
+            .ok_or(StopMicrovmError::GuestMemoryNotInitialized)?;
+
+        self.device_manager
+            .remove_devices(
+                vm_as.clone(),
+                self.epoll_manager.clone(),
+                self.address_space.address_space(),
+            )
+            .map_err(StopMicrovmError::DeviceManager)
+    }
+
+    /// Reset the console into canonical mode.
+    pub fn reset_console(&self) -> std::result::Result<(), DeviceMgrError> {
+        self.device_manager.reset_console()
+    }
+
+    pub(crate) fn init_dmesg_logger(&mut self) {
+        let writer = self.dmesg_logger();
+        self.dmesg_fifo = Some(writer);
+    }
+
+    /// dmesg write to logger
+    fn dmesg_logger(&self) -> Box<dyn io::Write + Send> {
+        Box::new(DmesgWriter::new(&self.logger))
+    }
+
+    pub(crate) fn init_guest_memory(&mut self) -> std::result::Result<(), StartMicroVmError> {
+        info!(self.logger, "VM: initializing guest memory...");
+        // We are not allowing reinitialization of vm guest memory.
+        if self.address_space.is_initialized() {
+            return Ok(());
+        }
+
+        // vcpu boot up require local memory. reserve 100 MiB memory
+        let mem_size = (self.vm_config.mem_size_mib as u64) << 20;
+
+        let mem_type = self.vm_config.mem_type.clone();
+        let mut mem_file_path = String::from("");
+        if mem_type == "hugetlbfs" {
+            let shared_info = self.shared_info.read()
+                .expect("Failed to determine if instance is initialized because shared info couldn't be read due to poisoned lock");
+            mem_file_path.push('/');
+            mem_file_path.push_str(shared_info.id.as_str());
```

Comment:
Should this path include something like `dragonball`?

Reply:
Changed.